wagtail/contrib/settings/context_processors.py

```diff
@@ -40,33 +38,46 @@ def __missing__(self, model_name):
         self[model_name] = value = self.get_setting(model_name)
         return value
 
+    def __str__(self):
+        return "SettingModuleProxy({0})".format(self.app_label)
+
     def get_setting(self, model_name):
         """
         Get a setting instance
         """
         Model = registry.get_by_natural_key(self.app_label, model_name)
         if Model is None:
-            return None
+            raise RuntimeError(
+                f"Could not find model matching `{self.app_label}.{model_name}`."
+            )
 
-        if isinstance(self.request_or_site, Site):
-            return Model.for_site(self.request_or_site)
-        # Utilises cached value on request if set
-        return Model.for_request(self.request_or_site)
+        if issubclass(Model, BaseGenericSetting):
+            return Model.get_or_create()
+        elif issubclass(Model, BaseSiteSetting):
+            if self.request_or_site is not None:
+                if isinstance(self.request_or_site, Site):
+                    return Model.for_site(self.request_or_site)
 
-    def __str__(self):
-        return "SettingsModuleProxy({0})".format(self.app_label)
+                # Utilises cached value on request if set
+                return Model.for_request(self.request_or_site)
 
+            raise RuntimeError(
+                "`request` not found in context and `use_default_site` not set."
+            )
```

Comment:
Maybe we should modify this message slightly to be something like:
```suggestion
            raise RuntimeError(
                "Site-specific settings cannot be identified because `request` is not available in the context and `use_default_site` is False."
            )
```